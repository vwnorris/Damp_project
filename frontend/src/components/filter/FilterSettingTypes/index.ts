export * from './FilterSettingsGenre';
export * from './FilterSettingsPrice';
export * from './FilterSettingsReleaseDate';
export * from './FilterSettingsAchivements';
export * from './FilterSettingsPopularTags';

export * from './types';