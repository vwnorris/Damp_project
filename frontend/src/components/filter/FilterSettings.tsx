import React from 'react';
import { Box } from '@mui/material';
import { FilterSettingsAchievements, FilterSettingsGenre, FilterSettingsPopularTags, FilterSettingsPrice, FilterSettingsReleaseDate } from './FilterSettingTypes';
import { IFilterData } from '../../types';

export const FilterSettings = (props: {filter: string, callBack: (filter: string, data: IFilterData) => void}) => {

    const handleChangeFilter = (data: IFilterData) => {
        props.callBack(props.filter, data);
    }

    return (
        <Box sx={{
            maxWidth: "350px",
            width: 0.9,
            minWidth: '250px',
            padding: 2,
            overflow: "hidden"
        }}>
            {props.filter === "Genre" && <FilterSettingsGenre setFilterData={handleChangeFilter}/>}
            {props.filter === "Price" && <FilterSettingsPrice setFilterData={handleChangeFilter}/>}
            {props.filter === "ReleaseDate" && <FilterSettingsReleaseDate setFilterData={handleChangeFilter}/>}
            {props.filter === "Achievements" && <FilterSettingsAchievements setFilterData={handleChangeFilter}/>}
            {props.filter === "Tags" && <FilterSettingsPopularTags setFilterData={handleChangeFilter}/>}
        </Box>
    )
}