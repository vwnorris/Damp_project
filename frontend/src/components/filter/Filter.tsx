import React, { useRef, useState } from 'react';
import { ExpandMore } from '@mui/icons-material';
import { Accordion, AccordionDetails, AccordionSummary, Box, Card, Chip, Popover, Typography } from '@mui/material';
import { observer } from 'mobx-react';
import { useStores } from '../../hooks';
import { FilterSettings } from './FilterSettings';
import { IFilterData } from '../../types';

export const Filter = observer(() => {
    const { store } = useStores();
    const [activeFilter, setActiveFilter] = useState<string>("");

    const [anchorEl, setAnchorEl] = useState<HTMLDivElement | null>(null);
    const anchorRef = useRef<HTMLDivElement>(null);
    const open = Boolean(anchorEl);

    const [filterInfo, setFilterInfo] = useState<HTMLElement | null>(null);
    const openFilterInfo = Boolean(filterInfo);    


    const addFilter = (filter: string, data: IFilterData) => {
        store.filterStore.addFilter(filter, data);
        handleClose();
    }

    const removeFilter = (filter: string) => {
        store.filterStore.removeFilter(filter);
    }

    const handleClick = (event: React.MouseEvent<HTMLDivElement>, filter: string) => {
        setActiveFilter(filter);
        setAnchorEl(anchorRef.current);
    }

    const handleClose = () => {
        setAnchorEl(null);
        setActiveFilter("");
    }

    const handlePopoverOpen = (event: React.MouseEvent<HTMLElement>, filter: any) => {
        store.filterStore.selectFilter(filter)
        setFilterInfo(event.currentTarget);
      };
    
      const handlePopoverClose = () => {
        store.filterStore.unSelectFilter()
        setFilterInfo(null);
      };

    return (
        <Box>
           <Accordion data-testid="filters">
                <AccordionSummary expandIcon={<ExpandMore/>}>
                    <Box sx={{
                        display: 'flex',
                        flexDirection: 'row',
                        flexWrap: 'wrap',
                        justifyContent: 'center',
                        alignItems: 'center'
                    }}>
                        <Typography variant='h5'>Filters</Typography> 
                        {Array.from(store.filterStore.activeFilters).map(a => <Chip data-testid={a.name+"Active"} sx={{margin: 0.5, backgroundColor: a.color, color: 'white'}} key={`${a.name}-activeDisplay`} label={a.name}/>)}
                    </Box>
                </AccordionSummary>
                <AccordionDetails>
                <Typography variant='h6'>All filters</Typography>
                    <Card sx={{
                        display: 'flex',
                        flexDirection: 'row',
                        flexWrap: 'wrap',
                        padding: 2
                    }} ref={anchorRef}>
                        {Array.from(store.filterStore.allFilters).map(a => 
                            <Chip 
                                sx={{margin: 0.5, backgroundColor: a.color, color: 'white'}} 
                                key={`${a.name}`} 
                                label={a.name} 
                                variant='filled' 
                                clickable
                                data-testid={a.name} 
                                onClick={(event) => handleClick(event, a.name)}
                                aria-owns={open ? 'mouse-over-popover' : undefined}
                                aria-haspopup="true"
                                onMouseEnter={(event) => handlePopoverOpen(event, a)}
                                onMouseLeave={handlePopoverClose}/>
                        )}
                    </Card>
                    <Typography variant='h6'>Active filters</Typography>
                    <Card sx={{
                        display: 'flex',
                        flexDirection: 'row',
                        flexWrap: 'wrap',
                        padding: 2
                    }}>
                        {Array.from(store.filterStore.activeFilters).length === 0 && <Chip label={"No filters added"}/>}
                        {Array.from(store.filterStore.activeFilters).map(a => <Chip sx={{margin: 0.5, backgroundColor: a.color, color: 'white'}} key={`${a.name}-active`} label={`${a.name} | ${a.data!.visualData}`} clickable onClick={() => removeFilter(a.name)}/>)}
                    </Card>
                    <Popover open={open} onClose={handleClose} anchorEl={anchorEl} anchorOrigin={{vertical: 'bottom', horizontal: 'left'}}>
                        <FilterSettings filter={activeFilter} callBack={addFilter}/>
                    </Popover>
                    <Popover
                        id="mouse-over-popover"
                        sx={{ pointerEvents: 'none' }}
                        open={openFilterInfo}
                        anchorEl={filterInfo}
                        anchorOrigin={{ vertical: 'bottom', horizontal: 'center'}}
                        transformOrigin={{ vertical: 'top', horizontal: 'center'}}
                        onClose={handlePopoverClose}
                        disableRestoreFocus
                    >
                        <Typography variant='body2' sx={{p:2}}>{store.filterStore.selectedFilter?.description}</Typography>
                    </Popover>
                </AccordionDetails>
           </Accordion>
        </Box>
    )
});