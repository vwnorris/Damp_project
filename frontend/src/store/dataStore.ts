import { action, computed, IObservableArray, makeObservable, observable, runInAction } from "mobx";
import { Game, SortType } from "../types";
import { LOAD_GAMES } from "../graphQL";
import { client } from "../util";
import { ApolloClient, ApolloError, NormalizedCacheObject, ObservableQuery } from "@apollo/client";
import { RootStore } from "./rootStore";

interface ISort {
    ascending: boolean;
    type: SortType;
}

/**
 * Store handling data fetching and querying the database
 */
export class DataStore {
    data: IObservableArray<Game>;
    pageNumber: number;
    searchString: string = "";
    loading: boolean;
    allFound: boolean = false;
    error: ApolloError | undefined;
    sort: ISort;
    rootStore: RootStore
    query: ObservableQuery;

    constructor(rootStore: RootStore, inClient: ApolloClient<NormalizedCacheObject> = client) {
        this.query = inClient.watchQuery({query: LOAD_GAMES});
        this.pageNumber = 0
        this.loading = false;
        this.error = undefined;
        this.data = observable.array<Game>([]);
        this.sort = {
            ascending: true,
            type: SortType.NONE
        };
        this.rootStore = rootStore;
        makeObservable(this, {
            data: observable,
            loading: observable,
            error: observable,
            allFound: observable,
            searchString: observable,
            sort: observable,
            getMoreData: action,
            reloadData: action,
            setSort: action,
            setSearchString: action,
            resetStore: action,
            resetable: computed
        });
        runInAction(() => {
            this.getMoreData();
        });
    }

    get resetable(): boolean {
        if (this.pageNumber > 1 || this.data.length > 15 || this.sort.type !== SortType.NONE) return true
        return false;
    }

    async reloadData() {
        runInAction(() => {
            this.pageNumber = 0;
            this.data.clear();
            this.allFound = false;
            this.getMoreData();
        });
    }

    setSearchString(s: string) {
        this.searchString = s;
    }

    async getMoreData() {
        const variables = {
            page: this.pageNumber,
            filter: this.rootStore.filterStore.queryFilter,
            sort: this.sort,
            search: this.searchString
        };
        this.error = undefined;
        this.loading = true;
        this.query.refetch(
            variables
        ).then(
            action("getSucces", ({data}) => {
                this.data.push(...data.games);
                this.loading = false;
                this.pageNumber++;
                if (data.games.length < 15) {
                    this.allFound = true;
                }
            })
        ).catch(action("getError", (error: ApolloError) => {
            this.error = error;
        }))
    }

    setSort(type: SortType, ascending: boolean = true) {
        this.sort.ascending = ascending;
        this.sort.type = type;
    }

    resetStore() {
        this.pageNumber = 0
        this.loading = false;
        this.error = undefined;
        this.data.clear();
        this.sort = {
            ascending: true,
            type: SortType.NONE
        };
    }

}