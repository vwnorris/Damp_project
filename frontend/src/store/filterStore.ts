import { observable, action, makeObservable, runInAction, computed } from 'mobx';
import { IFilter, IFilterData } from '../types';
import { RootStore } from './rootStore';

interface ReturnData {
    genre?: string[];
    price?: number[];
    releasedate?: (Date | null)[];
    achivements?: number;
    tags?: string[];
}

/**
 * Store that handles filters and lets the data store get a query to send to the database
 */
export class FilterStore {
    // used sets here just because we want only unique values
    activeFilters: Set<IFilter> = new Set<IFilter>();
    allFilters: Set<IFilter> = new Set<IFilter>([
        {name: "Genre", color: "#F2994A", description: "The genre of the game"}, 
        {name: "Price", color: "#F2C94C", description: "The price of the game"}, 
        {name: 'ReleaseDate', color: '#BB6BD9', description: "The release date of the game"}, 
        {name: "Achievements", color: "#2D9CDB", description: "The number of achievements of the game"}, 
        {name: "Tags", color: "#EB5757", description: "The tags of the game"}]);
    queryFilter: ReturnData = observable.object<ReturnData>({});
    rootStore: RootStore;
    selectedFilter: IFilter | undefined;
    showModalInfo: boolean = false;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;
        makeObservable(this, {
            activeFilters: observable,
            allFilters: observable,
            queryFilter: observable,
            addFilter: action,
            removeFilter: action,
            resetStore: action,
            resetable: computed
        });
    }

    get resetable(): boolean {
        if (Array.from(this.activeFilters).length > 0) return true;
        return false;
    }

    /**
     * Adds a filter to the query when a filter is added such that the data store can use it to make a query with the new parameters
     */
    private addFilterQuery (filter: IFilter) {
        switch (filter.name) {
            case "Genre":
                this.queryFilter.genre = (filter.data!.data as string[]);
                break;
            case "Price":
                this.queryFilter.price = (filter.data!.data as number[]);
                break;
            case "ReleaseDate":
                this.queryFilter.releasedate = (filter.data!.data as (Date | null)[]);
                break;
            case "Achievements":
                this.queryFilter.achivements = (filter.data!.data as number);
                break;
            case "Tags":
                this.queryFilter.tags = (filter.data!.data as string[]);
                break;
        }
    }

    private removeFilterQuery (filter: IFilter) {
        type objectKey = keyof ReturnData;
        delete this.queryFilter[filter.name.toLowerCase() as objectKey];
    }

    private getFromSet(set: Set<IFilter>, name: string): IFilter {
        let elem: IFilter;
        for(let a of Array.from(set)) {
            if (a.name === name) {
                elem = a;
                break;
            }
        }
        set.delete(elem!);
        return elem!;
    }

    addFilter(filter: string, data: IFilterData) {
        const f = this.getFromSet(this.allFilters, filter);
        f.data = data;
        this.addFilterQuery(f);
        this.activeFilters.add(f);
        runInAction(() => {
            this.rootStore.dataStore.reloadData();
        });
    }

    removeFilter(filter: string) {
        const f = this.getFromSet(this.activeFilters, filter);
        this.removeFilterQuery(f);
        this.allFilters.add(f);
        runInAction(() => {
            this.rootStore.dataStore.reloadData();
        });
    }

    selectFilter(filter: IFilter) {
        this.selectedFilter = filter;
        this.showModalInfo = true;
    }

    unSelectFilter() {
        runInAction(() => {
            this.showModalInfo = false;
            this.selectedFilter = undefined;
        });
    }

    resetStore() {
        Array.from(this.activeFilters).forEach(filter => {
            const f = this.getFromSet(this.activeFilters, filter.name);
            this.removeFilterQuery(f);
            this.allFilters.add(f);
        });
    }
}