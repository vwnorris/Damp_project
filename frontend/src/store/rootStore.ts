import { ApolloClient, NormalizedCacheObject } from "@apollo/client";
import { action, computed, makeObservable } from "mobx";
import { client } from "../util";
import { DataStore } from "./dataStore";
import { FilterStore } from "./filterStore";
import { ModalStore } from "./modalStore";

/**
 * This class is the backbone of our store system and lets the other stores communicate.
 * It is created with mobx and has our *database* as the default client
 */
export class RootStore {
    filterStore: FilterStore;
    dataStore: DataStore;
    modalStore: ModalStore;

    constructor(inClient: ApolloClient<NormalizedCacheObject> = client) {
        this.filterStore = new FilterStore(this);
        this.dataStore = new DataStore(this, inClient);
        this.modalStore = new ModalStore(this, inClient);
        makeObservable(this, {
            enableResetButton: computed,
            resetStores: action
        });     
    }

    /** Function to see if the reset button should come or not */ 
    get enableResetButton(): boolean {
        if (this.dataStore.resetable || this.modalStore.resetable || this.filterStore.resetable) return true;
        return false;
    }

    /** Function to reset all the stores */ 
    public resetStores(reloadData: boolean = true) {
        this.filterStore.resetStore();
        this.dataStore.resetStore();
        this.modalStore.resetStore();
        if (reloadData) this.dataStore.reloadData();
    }
}