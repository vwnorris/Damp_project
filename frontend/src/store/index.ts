import { createContext } from "react";
import { RootStore } from "./rootStore";

export const store = new RootStore()

export const defaultContext = createContext({
    store: store
});
