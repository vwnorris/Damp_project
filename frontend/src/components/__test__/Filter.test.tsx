import { Filter } from "../filter";
import { render, screen, cleanup, within } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { Provider } from "mobx-react";
import { defaultContext, store } from "../../store";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";

beforeEach(() => {
  store.resetStores();
});

afterEach(cleanup);

describe("Filter snapshot test", () => {
  it("render filter component", () => {
    const tree = renderer.create(<Filter />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Filter test", () => {
  it("renders without crashing", () => {
    render(<Filter />);
  });

  it("clicking filters display filtercategories", () => {
    render(<Filter />);
    expect(screen.getByRole("button", { name: /filters/i })).toBeEnabled();
    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    screen.getByText(/all filters/i);

    screen.getByText(/genre/i);
    screen.getByText(/price/i);
    screen.getByText(/releasedate/i);
    screen.getByText(/achievements/i);
    screen.getByText(/tags/i);

    screen.getByText(/no filters added/i);
  });

  it("clicking filtercategory shows correct options, and is clickable", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/genre/i).click();
    });

    screen.getByText(/valve/i);
    screen.getByText(/gore/i);
    screen.getByText(/rpg/i);
    screen.getByText(/action/i);
  });

  it("renders checkboxes in genre correctly, and can be checked", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/genre/i).click();
    });

    expect(screen.getByRole("checkbox", { name: /valve/i })).not.toBeChecked();
    expect(screen.getByRole("checkbox", { name: /gore/i })).not.toBeChecked();
    expect(screen.getByRole("checkbox", { name: /rpg/i })).not.toBeChecked();
    expect(screen.getByRole("checkbox", { name: /action/i })).not.toBeChecked();

    act(() => {
      screen
        .getByRole("checkbox", {
          name: /valve/i,
        })
        .click();
      screen
        .getByRole("checkbox", {
          name: /gore/i,
        })
        .click();
      screen
        .getByRole("checkbox", {
          name: /rpg/i,
        })
        .click();
      screen
        .getByRole("checkbox", {
          name: /action/i,
        })
        .click();
    });

    expect(screen.getByRole("checkbox", { name: /valve/i })).toBeChecked();
    expect(screen.getByRole("checkbox", { name: /gore/i })).toBeChecked();
    expect(screen.getByRole("checkbox", { name: /rpg/i })).toBeChecked();
    expect(screen.getByRole("checkbox", { name: /action/i })).toBeChecked();
  });

  it("test hover functions on filter buttons", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    userEvent.hover(screen.getByRole('button', {
      name: /genre/i,
      hidden: true
    }))

    screen.getByText(/the genre of the game/i)

    userEvent.hover(screen.getByRole('button', {
      name: /price/i,
      hidden: true
    }))

    screen.getByText(/the price of the game/i)

    userEvent.hover(screen.getByRole('button', {
      name: /tags/i,
      hidden: true
    }))

    screen.getByText(/the tags of the game/i)

    userEvent.click(screen.getByRole('button', {
      name: /tags/i,
      hidden: true
    }))
  });

  it("render filterbuttons correctly", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/genre/i).click();
    });

    act(() => {
      screen
        .getByRole("checkbox", {
          name: /valve/i,
        })
        .click();
    });
  });

  it("can apply filter", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/achievements/i).click();
    });

    act(() => {
      screen
        .getByRole("radio", {
          name: /has achievements/i,
        })
        .click();
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    const button = screen.getByRole("button", {
      name: /filters achievements/i,
    });

    within(button).getByText(/achievements/i);

    act(() => {
      screen
        .getByRole("button", {
          name: /achievements \| has achievements/i,
        })
        .click();
    });
  });

  it("store testing", () => {
    render(
      <Provider {...defaultContext}>
        <Filter />
      </Provider>
    );

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/achievements/i).click();
    });

    act(() => {
      screen
        .getByRole("radio", {
          name: /has achievements/i,
        })
        .click();
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    const button = screen.getByRole("button", {
      name: /filters achievements/i,
    });

    within(button).getByText(/achievements/i);

    expect(store.filterStore.activeFilters.size).toEqual(1);

    store.filterStore.activeFilters.forEach((filter) => {
      expect(filter.name).toMatch(/achievements/i);
      expect(filter.data?.visualData).toMatch(/has achievements/i);
    });

    act(() => {
      screen.getByText(/genre/i).click();
    });

    act(() => {
      screen
        .getByRole("checkbox", {
          name: /action/i,
        })
        .click();
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    expect(store.filterStore.activeFilters.size).toEqual(2);

    store.filterStore.activeFilters.forEach((filter) => {
      expect(filter.name).toMatch(/achievements|genre/i);
      expect(filter.data?.visualData).toMatch(/has achievements|action/i);
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /achievements \| has achievements/i,
        })
        .click();
    });

    expect(store.filterStore.activeFilters.size).toEqual(1);

    store.filterStore.activeFilters.forEach((filter) => {
      expect(filter.name).toMatch(/genre/i);
      expect(filter.data?.visualData).toMatch(/action/i);
    });
  });

  it("apply price filter", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/price/i).click();
    });

    act(() => {
      screen.getByTestId("slider").click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });
  });

  it("apply date filter with invalid date", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/releasedate/i).click();
    });

    userEvent.type(
      screen.getByRole("textbox", {
        name: /after/i,
      }),
      "23"
    );

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/releasedate \| after invalid date/i);
  });

  it("apply date filter with valid start and end date", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/releasedate/i).click();
    });

    userEvent.type(
      screen.getByRole("textbox", {
        name: /after/i,
      }),
      "06062022"
    );

    userEvent.type(
      screen.getByRole("textbox", {
        name: /before/i,
      }),
      "07062022"
    );

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/releasedate \| between 06\/06\/2022 and 06\/07\/2022/i);
  });

  it("apply date filter with valid start", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/releasedate/i).click();
    });

    userEvent.type(
      screen.getByRole("textbox", {
        name: /before/i,
      }),
      "07062022"
    );

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/releasedate \| before 06\/07\/2022/i);
  });

  it("apply tags", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/tags/i).click();
    });

    act(() => {
      screen
        .getByRole("combobox", {
          name: /tags/i,
        })
        .click();
    });

    userEvent.type(
      screen.getByRole("combobox", {
        name: /tags/i,
      }),
      "2d{arrowdown}{enter}"
    );

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });
  });

  it("apply achievements filter without choosing an option", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/achievements/i).click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/please select a value to filter/i);
  });

  it("apply genre filter without choosing an option", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/genre/i).click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/please select genre\(s\) to filter/i);
  });

  it("apply tags filter without choosing an option", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/tags/i).click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/please select a tag to filter/i);
  });

  it("apply price filter without choosing an option", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/price/i).click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/no reason to filter between the min and max/i);
  });

  it("apply date filter without choosing an option", () => {
    render(<Filter />);

    act(() => {
      screen.getByRole("button", { name: /filters/i }).click();
    });

    act(() => {
      screen.getByText(/releasedate/i).click();
    });

    act(() => {
      screen
        .getByRole("button", {
          name: /add filter/i,
        })
        .click();
    });

    screen.getByText(/please select valid dates./i);
  });
});
