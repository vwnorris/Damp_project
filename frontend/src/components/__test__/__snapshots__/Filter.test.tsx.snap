// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter snapshot test render filter component 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
    data-testid="filters"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <div
          className="MuiBox-root css-1xmcsij"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Filters
          </h5>
        </div>
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            className="MuiAccordion-region"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <h6
                className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                All filters
              </h6>
              <div
                className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-100nkij-MuiPaper-root-MuiCard-root"
              >
                <div
                  aria-haspopup="true"
                  className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-10m89vb-MuiButtonBase-root-MuiChip-root"
                  data-testid="Genre"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    Genre
                  </span>
                </div>
                <div
                  aria-haspopup="true"
                  className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-1uf3vez-MuiButtonBase-root-MuiChip-root"
                  data-testid="Price"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    Price
                  </span>
                </div>
                <div
                  aria-haspopup="true"
                  className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-co3evw-MuiButtonBase-root-MuiChip-root"
                  data-testid="ReleaseDate"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    ReleaseDate
                  </span>
                </div>
                <div
                  aria-haspopup="true"
                  className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13pma9r-MuiButtonBase-root-MuiChip-root"
                  data-testid="Achievements"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    Achievements
                  </span>
                </div>
                <div
                  aria-haspopup="true"
                  className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-1y9da21-MuiButtonBase-root-MuiChip-root"
                  data-testid="Tags"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    Tags
                  </span>
                </div>
              </div>
              <h6
                className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                Active filters
              </h6>
              <div
                className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-100nkij-MuiPaper-root-MuiCard-root"
              >
                <div
                  className="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-gac2fo-MuiChip-root"
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                >
                  <span
                    className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                  >
                    No filters added
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
